### PR TITLE
chore: allow aria-allowed-role to run virtually

### PR DESCRIPTION
**Pull request by straker**
Converting `is-visible` was a bit of a pain...

Part of issue #2734

### CODE REVIEW

File: lib/commons/aria/get-element-unallowed-roles.js [hunk 1]
```diff
@@ -25,26 +31,18 @@ const dpubRoles = [
...
   }
 
-  if (node.hasAttribute('role')) {
-    const nodeRoles = tokenList(node.getAttribute('role').toLowerCase());
+  if (vNode.hasAttr('role')) {
+    const nodeRoles = tokenList(vNode.attr('role').toLowerCase());
     roles = roles.concat(nodeRoles);
   }
 
-  if (node.hasAttributeNS('http://www.idpf.org/2007/ops', 'type')) {
```
Comment (straker) [inline lib/commons/aria/get-element-unallowed-roles.js:40]: We determined that we no longer need to support this so we're good to move out of it.

File: lib/commons/dom/is-visible.js [hunk 2]
```diff
@@ -112,50 +113,74 @@ function isVisible(el, screenReader, recursed) {
     );
   }
 
-  const vNode = getNodeFromTree(el);
+  const vNode = el instanceof AbstractVirtualNode ? el : getNodeFromTree(el);
```
Comment (straker) [inline lib/commons/dom/is-visible.js:116]: The general strategy is check things that virtual node has access to, followed by things only a true DOM node has access to. A detatched virtual node is visible.

File: test/commons/dom/is-visible.js [hunk 3]
```diff
@@ -5,25 +5,24 @@ describe('dom.isVisible', function() {
...
   describe('default usage', function() {
     // Firefox returns `null` if accessed inside a hidden iframe
     it('should return false if computedStyle return null for whatever reason', function() {
-      var orig = window.getComputedStyle;
-      window.getComputedStyle = function() {
-        return null;
-      };
-      assert.isFalse(axe.commons.dom.isVisible(fakeNode));
-      window.getComputedStyle = orig;
+      computedStyleStub = sinon.stub(window, 'getComputedStyle').returns(null);
```
Comment (straker) [inline test/commons/dom/is-visible.js:23]: If this test failed it would not reset the override and so all future tests would fail.

File: lib/core/utils/is-html-element.js [hunk 4]
```diff
@@ -1,137 +1,20 @@
...
 /**
  * Verifies that if a given html tag is valid
  * @method isHtmlElement
  * @memberof axe.utils
- * @param htmlTag htmlTag to check if valid
+ * @param {HTMLElement|VirtualNode} node node to check if valid
  * @return {Boolean} true/ false
  */
 function isHtmlElement(node) {
+  const nodeName = node.props?.nodeName ?? node.nodeName.toLowerCase();
```
Comment (WilcoFiers) [inline lib/core/utils/is-html-element.js:11]: <3

File: lib/core/utils/is-html-element.js [hunk 5]
```diff
@@ -1,137 +1,20 @@
...
+
   if (node.namespaceURI === 'http://www.w3.org/2000/svg') {
     return false;
   }
 
-  return htmlTags.includes(node.nodeName.toLowerCase());
+  return Object.keys(standards.htmlElms).includes(nodeName);
```
Comment (WilcoFiers) [inline lib/core/utils/is-html-element.js:17]: Don't think you need to create an array here:

```suggestion
  return !!standards.htmlElms[nodeName];
```

The other thing about this, if someone for some reason wanted to unset an HTML element with `axe.configure({ standards })`, they they wouldn't be able to, since this still returns true, even if the prop on `htmlElms` is undefined or null.

File: lib/commons/aria/get-element-unallowed-roles.js [hunk 6]
```diff
@@ -54,21 +52,22 @@ function getRoleSegments(node) {
 /**
  * gets all unallowed roles for a given node
  * @method getElementUnallowedRoles
- * @param {Object} node HTMLElement to validate
- * @param {String} tagName tag name of a node
+ * @param {Object|VirtualNode} node HTMLElement to validate
```
Comment (WilcoFiers) [inline lib/commons/aria/get-element-unallowed-roles.js:55]:
```suggestion
 * @param {Element|VirtualNode} node HTMLElement to validate
```

File: test/checks/aria/aria-allowed-role.js [hunk 7]
```diff
@@ -1,421 +1,333 @@
...
     var actual = axe.testUtils
       .getCheckEvaluate('aria-allowed-role')
-      .call(checkContext, target, options);
+      .call(checkContext, null, options, vNode);
     var expected = false;
     assert.equal(actual, expected);
     assert.deepEqual(checkContext._data, ['row']);
   });
 
-  it('returns true when A has namespace as svg', function() {
```
Comment (WilcoFiers) [inline test/checks/aria/aria-allowed-role.js:45]: Where did this go?

File: test/checks/aria/aria-allowed-role.js [hunk 8]
```diff
@@ -1,421 +1,333 @@
...
     );
     assert.deepEqual(checkContext._data, ['none']);
   });
 
   it('returns true when input of type image and no role', function() {
-    var node = document.createElement('img');
-    node.setAttribute('type', 'image');
-    fixture.appendChild(node);
-    flatTreeSetup(fixture);
+    var vNode = queryFixture('<img id="target" type="image"/>');
```
Comment (WilcoFiers) [inline test/checks/aria/aria-allowed-role.js:139]:
```suggestion
    var vNode = queryFixture('<input id="target" type="image"/>');
```

File: test/commons/aria/get-element-unallowed-roles.js [hunk 9]
```diff
@@ -66,16 +67,28 @@ describe('aria.getElementUnallowedRoles', function() {
...
     var node = document.createElement('tr');
     node.setAttribute('role', 'row');
     flatTreeSetup(node);
-    var actual = axe.commons.aria.getElementUnallowedRoles(node, false);
+    var actual = getElementUnallowedRoles(node, false);
     assert.isNotEmpty(actual);
     assert.include(actual, 'row');
   });
+
+  it('works with SerialVirtualNode', function() {
```
Comment (WilcoFiers) [inline test/commons/aria/get-element-unallowed-roles.js:83]: Can you add at least one additional test showing it doesn't always return empty?

File: test/rule-matches/aria-allowed-role-matches.js [hunk 10]
```diff
@@ -1,47 +1,27 @@
...
+  var queryFixture = axe.testUtils.queryFixture;
   var rule;
 
   beforeEach(function() {
     rule = axe.utils.getRule('aria-allowed-role');
   });
 
-  afterEach(function() {
-    fixture.innerHTML = '';
-  });
```
Comment (WilcoFiers) [inline test/rule-matches/aria-allowed-role-matches.js:14]: Why did you remove this?

File: test/integration/virtual-rules/aria-allowed-role.js [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+        role: 'link'
+      }
+    });
+
+    assert.lengthOf(results.passes, 1);
+    assert.lengthOf(results.violations, 0);
+    assert.lengthOf(results.incomplete, 0);
+  });
+
+  it('should incomplete for hidden element', function() {
```
Comment (WilcoFiers) [inline test/integration/virtual-rules/aria-allowed-role.js:70]: Why is this incomplete?

**Review by WilcoFiers — CHANGES_REQUESTED** [review]

**Comment by WilcoFiers** [comment]
Also, change the PR title

File: test/checks/aria/aria-allowed-role.js (hunk 7 above)
Comment (straker) [reply test/checks/aria/aria-allowed-role.js:45]: https://github.com/dequelabs/axe-core/pull/3067#discussion_r665746595

File: test/rule-matches/aria-allowed-role-matches.js (hunk 10 above)
Comment (straker) [reply test/rule-matches/aria-allowed-role-matches.js:14]: Our testutils file does this for us https://github.com/dequelabs/axe-core/blob/develop/test/testutils.js#L440-L442

File: test/integration/virtual-rules/aria-allowed-role.js (hunk 11 above)
Comment (straker) [reply test/integration/virtual-rules/aria-allowed-role.js:70]: Because hidden elements with unallowed roles return undefined https://github.com/dequelabs/axe-core/blob/develop/lib/checks/aria/aria-allowed-role-evaluate.js#L33-L35

**Review by WilcoFiers — APPROVED** [review]
